### ISSUE TITLE
Error on Review Page - Submit Button 

### ISSUE THREAD

**Issue by jonathanjnelson** [department-of-veterans-affairs/va.gov-team#11594]
## Issue Description
When a user attempts to submit a form on VA.gov and receives an error message on the review page, the submit button  disappears. 

1) what is the current behavior?
2) what is the intended/desired behavior?
3) what needs to happen to get from A to B?

Note: Mapped the behavior for the 10-10CG [here](https://github.com/department-of-veterans-affairs/va.gov-team/issues/12179)

## Current behavior on VA.gov forms

-  **Submit button becomes inactive** - If you successfully submit and use the browser back button to return to the review page 
- **Submit button disappears** -  When there is an error (in production) 
- **Submit button reappears** -  You don't have to modify anything, just save the form in progress and reload is usually enough for it to reappear 

## Tasks
- [ ] Align with development and business on intended behavior 
- [ ] Discover why submit button disappears on some forms but not others 
- [ ] Address what needs to happen to make necessarily changes 

## Acceptance Criteria
- [ ] Understanding of the submit button when an error message is triggered
---
## How to configure this issue
- [x] **Attached to a Milestone** (when will this be completed?)
- [x] **Attached to an Epic** (what body of work is this a part of?)
- [x] **Labeled with Team** (`product support`, `analytics-insights`, `operations`, `service-design`, `tools-be`, `tools-fe`)
- [x] **Labeled with Practice Area** (`backend`, `frontend`, `devops`, `design`, `research`, `product`, `ia`, `qa`, `analytics`, `contact center`, `research`, `accessibility`, `content`)
- [x] **Labeled with Type** (`bug`, `request`, `discovery`, `documentation`, etc.)

**Comment by jonathanjnelson** [comment]
From @Clark-W-Griswold and @jonathanjnelson conversations: 

- The submit button is the intended behavior in an effort to reduce excess api calls. 
- It seems like _sometimes_ the submit button remains (see 10-10EZ) where as the 10-10CG and 526 disappear 
- Why is this? From what Dan has seen in the code, it is based on the type of error received from the API. (throttled/client error/generic error/etc)

<img width="611" alt="Screen Shot 2020-08-03 at 2 43 27 PM" src="https://user-images.githubusercontent.com/58444931/89220784-cc255c80-d597-11ea-85a2-12a1f941d0bf.png">

**Comment by meganhkelley** [comment]
Hi @jonathanjnelson ! This behavior (bug?) is on VSP FE Tools team's (@Clark-W-Griswold) radar to dig into, but I hear you are actively working on it. Is your current goal to document existing behavior (including which forms this happens on)? We will hold off on doing work here until we hear from you — does that work for you? 

For what it's worth, @rianfowler has mentioned that this behavior is very possibly intentional to prevent users from submitting a series of erroneous submissions. Once current behavior is documented and we determine what desired behavior is, we will likely need to loop in VSP BE Tools team for awareness and approval. 

Let me know if you have questions!

**Comment by jonathanjnelson** [comment]
Hi @meganhkelley! 

I think there are two goals:

- To find out if there is any sort of workaround the Caregiver team can use in this sprint/or the next with the desire is to have submit button remain on the review page (if a error message is received) 
- Document intended and current behavior (with the understanding that this is the intended behavior - to prevent users from submitting a series of erroneous submissions)

**Comment by meganhkelley** [comment]
@jonathanjnelson are you working on #2?

**Comment by meganhkelley** [comment]
VSP FE Tools: 

- Document current behavior in code
- Discuss with VSP BE Team if there is rate limiting in forms library, why, etc
- Once we understand how this works the way it works and why, we can help Caregiver team

**Comment by jonathanjnelson** [comment]
Hi @meganhkelley - I think that plan sounds good! I do not have the expertise to document current behavior in code.

**Comment by meganhkelley** [comment]
Megan Kelley  34 minutes ago
hey @Jonathan Nelson sorry for the confusion here, my last comment was notes I took during a quick sync with VSP folks https://github.com/department-of-veterans-affairs/va.gov-team/issues/11594
to clarify, though, are you working on documenting the FE behavior on VA.gov (not in the code)? (edited) 

Alayna Abel  28 minutes ago
I can chime in here! No, the goal is not to document behavior across VA.gov. The goal is to understand that the current behavior is on Caregivers, and to find out if there are any tools available to us within the next sprint or so to change the behavior. And to generally keep on track of the work & investigation that VSP is doing, in case those tools become available later on. Does that make sense? (edited) 

Jonathan Nelson  25 minutes ago
:this: I am sorry for any confusion @Megan Kelley

Megan Kelley  3 minutes ago
OhHHHHHHhHhHHHHhhhh hoooooooo ok I understand! I left a comment with VSP's approach. we are going to look at this this sprint (our sprints are different than yours) and will provide updates on the ticket.
I can't make any promises on what behavior we can set up for y'all in the short term, but I do hope to have an update for you early/mid next week!
New

Alayna Abel  < 1 minute ago
Perfect, thank you!!

**Comment by Clark-W-Griswold** [comment]
When submission status is equal to the below cases, we show a submit button:
`false`
`submitPending - button is disabled`
`applicationSubmitted`
`clientError`
`throttledError`
`validationError`

And the catch all case when it is NOT production will have a submit button.

**Comment by rianfowler** [comment]
@Clark-W-Griswold what determines the submission status? is this a enum in the redux state or a status returned from the backend?

**Comment by Clark-W-Griswold** [comment]
These are the values as used on the frontend. Those values are mapped as follows:

```
false = default state. no status set.
submitPending - button is disabled = called once submit button has been clicked 
applicationSubmitted = set when a successful response has been received from the form submission

validationError = set if the form is submitted and has failed client side validation 
```

If the form submission to the api returns an error:
```
clientError = error message does not start with 'vets_throttled_error' or 'vets_server_error'
throttledError = error message starts with 'vets_throttled_error'
catch all status = error message starts with 'vets_throttled_error'
```

**Comment by rianfowler** [comment]
Which one of these statuses causes the button to disappear in general and which one causes it to disappear in this specific case?

**Comment by Clark-W-Griswold** [comment]
It's just our catch all in the case that the environment is production. Sorry for lack of clarity there. 

so if the following conditions are NOT met:
```
false
submitPending - button is disabled
applicationSubmitted
clientError
throttledError
validationError
```

Then we conditionally set the `submitButton` variable (which is the placeholder for the submit button) based on the following logic:
```
    if (process.env.NODE_ENV !== 'production') {
      submitButton = (
        <div className="small-6 usa-width-one-half medium-6 columns">
          <a onClick={onSubmit}>Submit again</a>
        </div>
      );
    }
```

**Comment by Clark-W-Griswold** [comment]
@jonathanjnelson given the above behavior of the submit buttons on the review page, is there an alternative expectation around when a submit button should/should not be displayed? Possibly a specific condition/response from the API would be subject to that case?

**Comment by sshein** [comment]
Just reading this thread. Can someone tell me in plain english what are the use cases currently when the button is not showing? I personally am confused about why we would ever want the submit button to disappear. Best practice for forms is to always show a button (not even have a disabled state), and rely on errors post-submit to tell the user what is going on.

**Comment by Clark-W-Griswold** [comment]
Hey @sshein! Here would be the flow from a user perspective:

- User lands on review page
- User clicks submit button
- Submit button transitions to disabled state
- Code submits data to the server
- ERROR from server received 
- This error does not map to a specific error state that we handle
- Default error state is loaded on the UI

If the environment is production:
- This default error state DOES NOT contain a submit button

If the environment is staging/local/etc:
- This default error state DOES contain a `submit again` link (not button, but text)

**Comment by sshein** [comment]
got it! thank you. I'm going to bring this up with the design system council tomorrow if there's time and get back with the recommendation. Can you snag a screencap of the default state in staging with the "submit again" link for me?

**Comment by jonathanjnelson** [comment]
Does this work, Shawna? 

<img width="509" alt="Screen Shot 2020-08-13 at 3 01 53 PM" src="https://user-images.githubusercontent.com/58444931/90181429-094ad500-dd76-11ea-93b1-a6062e46178b.png">

**Comment by sshein** [comment]
yes! thank you.

**Comment by cvalarida** [comment]
I've done some digging into this and have come to the same conclusion Dan has.

Currently, the submit button disappears after the form has been submitted and there was an error on the API side that _isn't_ handled by the client (such as a throttling error). I believe the historical context for this functionality goes way back. If memory serves, I think this happens because, when the API returns such an error, re-submitting the form won't change the result. It'll continue to fail until the problem is resolved on our end. Keeping the submit button there gives the veteran false hope that if they keep poking it, something might change.

To be clear, it's _possible_ that any given form will return an error that resolves itself after a little while, so poking that submit button a bunch of times _may_ actually be fruitful. Eventually. That _shouldn't_ be the case, but that really depends on how the endpoint was built, which isn't guaranteed to be consistent from form to form.

**Comment by meganhkelley** [comment]
This topic is on the agenda for design system council meeting 8/28

**Comment by KevinMHoffmanUSDS** [comment]
Good afternoon! 

The design system council has decided to change the behavior on this button in the following ways:
* Change the disappeared button to a disabled button. 
* Button becomes re-enabled upon editing anything on the review screen
* Accessibility suggestion: include a generic sentence OR content reference (to the error text below or above) to the error related to the disabling of the button

This is a globally recommended change.

**Comment by ala-yna** [comment]
Hi @KevinMHoffmanUSDS and @meganhkelley - thank you for tracking this! Just to follow up - can we get a ballpark of when VSP will be implementing the proposed changes? Thank you!

**Comment by meganhkelley** [comment]
This is slated for future work on VSP. @ala-yna please let me know if this becomes urgent for your team or is blocking you in any way. In the meantime, I'll move this ticket to `Ready`

**Comment by ala-yna** [comment]
Hi @caw310! Do you know what the status of this is?

**Comment by caw310** [comment]
@ala-yna , we haven't reviewed. Is this an urgent issue? Can review today.

**Comment by ala-yna** [comment]
Where can I go to see what you all are working on? Not a top priority, but want to make sure this is somewhere on the list.

**Comment by caw310** [comment]
You can sort on vsp-design-system-team and see what we are working on. As you will see we have a lot in the backlog and we're a team of 4 so we will get to it when there is capacity. If it is urgent, you can let me know and I'll work on prioritizing. Thanks.

**Comment by caw310** [comment]
Here is a link to our Zenboard.  https://app.zenhub.com/workspaces/vsp-design-system-5f8de67192551b0012ebb802/board?labels=design-system-team,vsp-design-system-team&filterLogic=any&useDefaultFilterLogic=false&repos=133843125,152664242,194202180

**Comment by mlighdc** [comment]
@caw310 can you please provide an status update? I was not able to find the ticket on the Zenboard. Thanks!

**Comment by taylorkaren** [comment]
With the current Forms Library being sunsetted, addressing this issue will not be prioritized. The focus of the Forms Library Team's work is on building out a better review page pattern in the new VA Forms System Core (VAFSC) package.

Addressing this specific issue will be part of the analysis work to be done on submitting form data to vets-api. [Analysis ticket for this work[(https://app.zenhub.com/workspaces/forms-library---platform-spike-team-61b0ae1f2cd3c30014e8a5b0/issues/department-of-veterans-affairs/va-forms-system-core/542) has been created; timeframe to work on this is TBD as of 8/15/22.

**Comment by caw310** [comment]
Closing as this is old there is a team focused on building forms and updating the forms library. If this is still an issue, file a new bug for the forms team.
